classes/form/CustomerAddressPersister.php

```diff
@@ -68,6 +68,11 @@ public function save(Address $address, $token)
             $old_address = new Address($address->id);
             $address->id = $address->id_address = null;
 
+            $this->cart->updateAddressId(
+                $old_address->id,
+                $address->id
```

Comment:
This value equals `null` with the previous line.

I guess we should save the address first, don't we?

Reply:
I don't see why to save it first?

Reply:
If you don't save it first, then the cart will be updated using `id_address_delivery = 0`

Reply:
Actually wouldn't it be more logical to first save the new address, so that a new id is created and **then** update the cart address?

Reply:
I made a change, please check.

Reply:
That's not what I meant. Saving the address at this moment won't change anything.
If what you are trying to do is update the Cart address id then you need both the former and the new id, so the address object must be saved to get its newly created ID, something like:

```php
        if ($address->isUsed()) {
            $old_address = new Address($address->id);
            $address->id = $address->id_address = null;

            if (!$address->save()) {
                return false;
            }

            //Now $address->id is not null any more
            $this->cart->updateAddressId(
                $old_address->id,
                $address->id
            );

            return $old_address->delete();
        }
```

Reply:
@jolelievre Thank you for clarifying, please check now.

Reply:
This seems more logical, let's see what the QA thinks about it ^^

Reply:
It gives error 500 in my stage env

Reply:
please enable debug mode and take a screenshot.

Reply:
Translation:
The property Address->id_country is empty
https://i.imgur.com/TvWMV00.jpg

Reply:
@simondaigre this can be preventted by changing `delete()` method in `Address.php`, change:
```        
} else {
            $this->deleted = true;
            return $this->update();
}
```

to:

```        
} else {
            $this->deleted = true;
            $this->setFieldsToUpdate(['delete']);
            return $this->update();
}
```